code/firmware/rosco_m68k_firmware/bootstrap.asm

```diff
@@ -211,6 +223,22 @@ EFP_DUMMY_NEGONE_D0L::
     rts
 
 
+; Initialize device blocks
+INITDEVS:
+    clr.w   DEVICE_COUNT    
+    move.w  #C_NUM_DEVICES,D0
+    lsl.w   #2,D0
+    lea.l   DEVICE_BLOCKS,A0
+    bra.s   .START
+
+.LOOP:
+    clr.l   (A0)+
+
+.START:
+    dbra.w  D0,.LOOP
+    rts

```

Comment:
Will this code clear the entire array? `DEVICE_BLOCKS` has a size of `C_DEVICE_SIZE * C_NUM_DEVICES = 0x20 * C_NUM_DEVICES`, but I think this will clear `((C_NUM_DEVICES << 2) + 1) * sizeof(long) = 0x10 * C_NUM_DEVICES + 4` bytes.

I haven't tested if this assembles (I know that some assemblers are picky about arithmetic) or works, but maybe something like this would work (though this requires a non-0 number of devices to clear):
```suggestion
INITDEVS:
    clr.w   DEVICE_COUNT    
    move.w  #C_DEVICE_SIZE*C_NUM_DEVICES-1,D0
    lea.l   DEVICE_BLOCKS,A0
    bra.s   .START

.LOOP:
    clr.b   (A0)+

.START:
    dbra.w  D0,.LOOP
    rts
```
Alternatively, adding a right-shift (or factoring out a multiple of 4) would allow the clear to still by long-wise, though it would require `C_DEVICE_SIZE` to be definitely divisible by 4.

Reply:
Hmm, yeah, I see what you mean, this will only clear half of the array (the left shift 2 is left over from before I increased the device size from 16 to 32). But I think just switching that to a left shift 3 will fix it - I don't see it "off by one" as I think you're suggesting (with the `-1`)?

I'll make the change now (and add a comments in the shared file about the size needing to be a multiple of 4) but happy to revisit if you still think it's wrong :)

Reply:
That looks good to me! :+1: Looking at it now, I have no idea why I thought the loop counter needed to be one less, I must have somehow ignored the branch to `.START` that was sitting right there. :sweat_smile: